### DISCUSSION TITLE
How to make remix-i18next work for all routes

### DISCUSSION THREAD

**Discussion by M4ss1ck** [remix-run/remix#2533] [Q&A]
Hi.

I'm trying to make a site based on Jokes App from docs. I added Tailwind CSS and `remix-i18next` but I'm unable to make my language switcher work, perhaps due to nested routes in Remix or my lack of React knowledge.

```
import styles from "~/styles/app.css";
import type { LinksFunction } from "remix";
import { Link } from "remix";

export const links: LinksFunction = () => {
  return [{ rel: "stylesheet", href: styles }];
};

function LanguageSwitcher() {
  return (
    <div className="flex flex-row items-center justify-center">
      <Link className="m-2" to="./?lng=es">
        es
      </Link>
      <Link className="m-2" to="./?lng=en">
        en
      </Link>
    </div>
  );
}
```

It's pretty simple and work for the root of the site, but it gets inconsistent in other routes. I placed it in a `Header.tsx` component.
I see that `<html lang="x">` changes but the content doesn't.

Is there any public repo that uses `remix-i18next` beyond `root.tsx`?
I would like to see more guides on i18n as I like Remix a lot ❤

FWIW [here](https://github.com/M4ss1ck/remix-music-genre-app) is my repo

**Comment by blasterbug** [comment]
I had the same issue and essentially solved it by putting the language in a cookie (the default language would be the user's browser language).  Then in `root.tsx` I added a loader that would read the cookie and load the right translation for the user, based on the value in the cookie or the default browser language. @sergiodxa had [this very good article](https://sergiodxa.com/articles/localizing-remix-apps-with-i18next "[Localizing Remix apps with i18next](https://sergiodxa.com/articles/localizing-remix-apps-with-i18next") on his blog about it. I just had to make some adjustment.

**Reply by brightpixels** [reply]
@blasterbug wouldn't detecting language in cookies require JS on client-side? I am asking because I am working on an app that has to work without JS and at the moment, when I toggle the language, it gets appended to the url `?lng=fr`. Is there a solution that don't rely on JS. (Update: I read the linked article and looks like request header is the best option in my use case. Looking for an example on how to create a route to trigger a request Accept Language header update without it being an actual route that can be reached by the user)

**Reply by sergiodxa** [reply]
You can read cookies server side

**Reply by momoirse** [reply]
I'm trying to implement internationalization (i18n) in my Remix Run application using the i18Next library. I want to have the locale specified in the URL, such as "es/home" for the Spanish version of the home page. I have the following code:

`
import i18n from "i18next";
import { useTranslation } from "react-i18next";
import Backend from "i18next-fs-backend";
import { initReactI18next } from "react-i18next";
import { useCallback } from "react";

export function useTranslationHelper(ns: string = "langs") {
  const { t, i18n } = useTranslation(ns);

  const translate = useCallback(
    (key: string) => {
      const fullKey = `${key}.${i18n.language}`;
      const translation = t(fullKey);
      // if translation includes .${i18n.language}, just return the key
      if (translation.includes(`.${i18n.language}`)) {
        return key;
      }
      return translation;
    },
    [t, i18n.language]
  );

  return translate;
}
`

However, I'm facing difficulties in setting up i18Next to work with the Remix Run framework and having the locale in the URL. I have already installed the necessary dependencies, including "i18next", "react-i18next", and "i18next-fs-backend".

Could someone please guide me on how to configure i18Next properly in Remix Run to support the desired URL structure with the locale?

Thank you in advance for your help!
